compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -12272,6 +12376,37 @@ private STNode parseOrderKeyListMemberEnd() {
         }
     }
 
+    private STNode parseGroupingKeyVariableDeclaration(boolean isRhsExpr) {
+        startContext(ParserRuleContext.GROUPING_KEY_LIST_VAR_DECLARATION);
+        STNode groupingKeyElementTypeDec =
+                parseTypeDescriptor(ParserRuleContext.TYPE_DESC_BEFORE_IDENTIFIER_IN_GROUPING_KEY);
+        STNode groupingKeyVarName = parseVariableName();
+        STNode equalsToken = parseAssignOp();
+        STNode groupingKeyExpression = parseExpression(OperatorPrecedence.QUERY, isRhsExpr, false);
+        endContext();
+        return STNodeFactory.createGroupingKeyVarDeclarationNode(groupingKeyElementTypeDec, groupingKeyVarName,
+                equalsToken, groupingKeyExpression);
+    }
+
+    /**
+     * Parse grouping key.
+     * <p>
+     * <code>grouping-key := variable-name | inferable-type-descriptor variable-name "=" expression</code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseGroupingKey(boolean isRhsExpr) {
+        STToken nextToken = peek();
+        SyntaxKind nextTokenKind = nextToken.kind;
+        if (nextTokenKind == SyntaxKind.IDENTIFIER_TOKEN) {
```

Comment:
An identifier can be a variable name or user-defined type. We need to resolve the ambiguity here